### PR TITLE
Exposed velocity on scroll position

### PR DESCRIPTION
**Pull request by pstromberg98**
## Description

This adds a getter to the `ScrollPosition` class which exposes the `ScrollActivity` velocity. As far as I am aware, there is currently no straight forward way to access the velocity. One use case for needing the velocity is making the padding or transforms of children of a scrollable change based off of the velocity (the referenced issue has a good example of what that could look like).

## Related Issues

This PR is in response to https://github.com/flutter/flutter/issues/26418

## Tests

Not sure what the test would look like for a simple getter

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by pstromberg98** [comment]
@Hixie I am not familiar with the all of the scroll internals so I hope this change is in the right place. Please let me know if that isn't the case.

The change is pretty simple so I am not sure if a test is needed. One test I could thing of is creating a ListView and then "catch" it mid scroll and ensure that the velocity matches the activity's velocity. Let me know if that is reasonable or necessary.

**Comment by Hixie** [comment]
The tests should verify whatever it is you would like us to make sure we don't revert. Imagine that after landing your PR, someone will write another PR that reverts everything except the tests, and will land as much of the revert as is possible without breaking the tests.

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 1]
```diff
@@ -106,6 +106,11 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
       restoreScrollOffset();
   }
 
+  /// If applicable, the velocity at which the scroll offset is currently
+  /// independently changing (i.e. without external stimuli such as a dragging gestures)
+  /// in logical pixels per second for the associated activity
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_position.dart:111]: Documentation must end with punctuation.

The docs should also say what happens if it's not applicable.
Also, what it means for it to be applicable.

Comment (Hixie) [reply packages/flutter/lib/src/widgets/scroll_position.dart:111]: Also we should say how to get notified when this changes.

Comment (pstromberg98) [reply packages/flutter/lib/src/widgets/scroll_position.dart:111]: @Hixie I took a lot of the documentation from the scroll activity velocity. I am wondering if (in addition to how to get notified of changes) the documentation for the scroll position velocity should be simplified to
`Represents the currently operative [ScrollActivity] velocity. If there is no current activity the velocity will be null.`.
If users want more details on velocity they could go look into ScrollActivity. That way we keep any mention of "applicable" or other specific vocab in the class to introduced it. Thoughts?

**Review by dnfield — COMMENTED** [review]
Rather than exposing new API, could something like the desired effect be achieved by using a custom `ScrollPhysics` that overrides the `createBallisticSimulation`?

**Comment by pstromberg98** [comment]
@dnfield If there is a way to do it with the existing mechanisms then I’d imagine that to be preferable. I’m not super familiar with the ScrollPhysics so I am going to learn about it. I’ll check out your suggestion and assuming that the effect is possible using a custom ScrollPhysics, I’ll create a package to easily achieve the effect.

I’ll close this issue assuming that things go well.

**Comment by pstromberg98** [comment]
@dnfield Okay so I have started poking around and the `createBallisticSimulation` is not enough to do what is needed (I think). If I understand correctly, the `createBallisticSimulation` is for creating simulations that are simulated after a user's drag ends? But the desired effect should occur when the drag ends *and* when the user is actively dragging which is why I think the `createBallisticSimulation` isn't enough. A could example of the effect is the iOS "Messages" app. 

But I also think that this PR wouldn't help achieve the effect either because when a user is actively dragging, the `DragScrollActivity` is used which has always has a velocity of 0 (let me know if I am incorrect there). I was under the assumption that there was always computed velocity based off of the offset change and time delta, which I don't think is the case. I apologize if I have gotten something wrong in my analysis here, I am just starting to look into the Flutter source so this is all pretty new to me.

All that being said I think the best solution is to calculate the velocity outside of Flutter by hooking into scroll changes as all we need is a previous offset, current offset, and the time delta.

**Comment by dnfield** [comment]
Yeah I think this makes sense.

Velocity only applies to ballistic simulations, not how fast the user is actually dragging.

That does seem like a useful thing to expose. We also might want to beef up the docs on velocity for future us :)

**Comment by pstromberg98** [comment]
https://github.com/pstromberg98/scroll_velocity_listener
Here is a link to the scroll velocity listener library (for posterity sake)

**Pull request closed without merging**